### DISCUSSION TITLE
[Question] MySQL plugin

### DISCUSSION THREAD

**Discussion by scw1004** [nanomq/nanomq#1796] [General]
我加了编译条件编译，设置了规则，发送了消息。但是不能保存到mysql的数据库里面。
用Mysql保存数据需要，除了需要建数据库，还需要建表和表头吗？

File: [/docs/zh_CN/rule/config-rule-engine.md](https://nanomq.io/docs/zh/latest/rule/config-rule-engine.html)
![1](https://github.com/nanomq/nanomq/assets/44936936/6b54d826-5e93-44e2-999e-1ccf9b6a427b)
![2](https://github.com/nanomq/nanomq/assets/44936936/c318084f-839c-4c80-b779-01fb27103b0b)

**Comment by JaylinYu** [comment]
是的

**Comment by scw1004** [comment]
请问一下，建表和表头的规则，要参考什么资料吗？看官网上没有说明的样子。

**Comment by StargazerWayne** [comment]
NanoMQ对表没有特殊的要求。可以根据应用场景需要，设计你自己的数据表，并在NanoMQ中进行相应的配置。

**Comment by scw1004** [comment]
> NanoMQ对表没有特殊的要求。可以根据应用场景需要，设计你自己的数据表，并在NanoMQ中进行相应的配置。

请问有没有相关的配置的例子，我看了官网找不到，官网上只有一个结果，没有配置的内容。

**Comment by StargazerWayne** [comment]
> 请问有没有相关的配置的例子，我看了官网找不到，官网上只有一个结果，没有配置的内容。

NanoMQ对MySQL的配置，你可以参考[文档](https://nanomq.io/docs/zh/latest/rule/config-rule-engine.html#mysql-%E8%A7%84%E5%88%99%E9%85%8D%E7%BD%AE)。文档中的注意点提到了，需要确保连接mysql时使用的各项正确，并确认你的database已经存在。 关于mysql的使用，这边没法给你详细的建议。

**Comment by JaylinYu** [comment]
MySQL 是个测试功能，只需要保证你写入的表存在应该就可以了。暂时不打算为mysql数据持久化提供例程，但此功能确实可用。

**Reply by huchangqiqi** [reply]
请教一个问题， rule_engine_insert_sql 函数里面写入数据库的时候，对于一个 rule 来说都是用的同一个全局的数据库连接，
例如 mysql_query 里面的 rules[i].mysql->conn ，如果并发写入是否有线程安全的问题。
我看默认是起了 16 个线程去从 taskq 里面取 task 然后执行 server_cb 的时候去调用的 rule_engine_insert_sql.
这个时候多个线程都去用的同一个数据库连接写入数据。
